src/protocol.rs

```diff
@@ -641,81 +555,153 @@ mod tests {
-        let presign_identifier = Identifier::random(&mut rng);
+        let QUORUM_SIZE = 3;
+        // Set GLOBAL config for participants
+        let configs = ParticipantConfig::random_quorum(QUORUM_SIZE, &mut rng);
+
+        // Set up auxinfo participants
+        let auxinfo_sid = Identifier::random(&mut rng);
+        let mut auxinfo_quorum = configs
+            .into_iter()
+            .map(|config| Participant::<AuxInfoParticipant>::from_config(&config, auxinfo_sid, ()))
+            .collect::<Vec<_>>();
+        let mut inboxes = HashMap::from_iter(
+            auxinfo_quorum
+                .iter()
+                .map(|p| (p.id, vec![]))
+                .collect::<Vec<_>>(),
+        );
+        let mut auxinfo_outputs: HashMap<
+            ParticipantIdentifier,
+            <AuxInfoParticipant as PublicProtocolParticipant>::Output,
+        > = HashMap::new();
 
-        for participant in &quorum {
+        // Initialize auxinfo for all parties
+        for participant in &auxinfo_quorum {
             let inbox = inboxes.get_mut(&participant.id).unwrap();
-            inbox.push(participant.initialize_auxinfo_message(auxinfo_identifier));
+            inbox.push(participant.initialize_message());
         }
 
-        while !is_auxinfo_done(&quorum, auxinfo_identifier)? {
-            process_messages(&mut quorum, &mut inboxes, &mut rng)?;
-        }
+        // Run auxinfo until all parties have outputs
+        while auxinfo_outputs.len() < QUORUM_SIZE {
+            let output = process_messages(&mut auxinfo_quorum, &mut inboxes, &mut rng)?;
 
-        for participant in &quorum {
-            let inbox = inboxes.get_mut(&participant.id).unwrap();
-            inbox.push(participant.initialize_keygen_message(keyshare_identifier));
-        }
-        while !is_keygen_done(&quorum, keyshare_identifier)? {
-            process_messages(&mut quorum, &mut inboxes, &mut rng)?;
+            if let Some((pid, output)) = output {
+                assert!(auxinfo_outputs.insert(pid, output).is_none());
+            }
         }
 
-        for participant in &mut quorum {
-            let message = participant.initialize_presign_message(
-                auxinfo_identifier,
-                keyshare_identifier,
-                presign_identifier,
-            )?;
+        // Auxinfo is done! Drop those participants
+        drop(auxinfo_quorum);
+
+        // Set up keygen participants
+        let keygen_sid = Identifier::random(&mut rng);
+        let configs = ParticipantConfig::random_quorum(QUORUM_SIZE, &mut rng);
+        let mut keygen_quorum = configs
+            .into_iter()
+            .map(|config| Participant::<KeygenParticipant>::from_config(&config, keygen_sid, ()))
+            .collect::<Vec<_>>();
+        let mut inboxes = HashMap::from_iter(
+            keygen_quorum
+                .iter()
+                .map(|p| (p.id, vec![]))
+                .collect::<Vec<_>>(),
+        );
```

Comment:
note: actually, we can probably reuse the inboxes in this case, since they should be empty after a successful protocol run.